modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/api/ast/ArtifactAstUtils.java

```diff
@@ -107,43 +147,106 @@ public static ArtifactAst parseAndBuildAppExtensionModel(String[] configResource
               .build()));
     } else {
-      logModelNotGenerated("Mule ExtensionModelLoader not found", muleContext);
+      logModelNotGenerated("Mule ExtensionModelLoader not found", artifactId, ast.getArtifactType());
       return empty();
     }
   }
 
+  /**
+   * Handles a {@link ValidationResult} uniformly. Warnings are logged, errors raise exceptions.
+   *
+   * @param validationResult The {@link ValidationResult} obtained using an {@link ArtifactAstValidator}.
+   * @throws ConfigurationException If there was any result item with {@link ERROR} level.
+   */
+  public static void handleValidationResult(ValidationResult validationResult)
+      throws ConfigurationException {
+    final Collection<ValidationResultItem> items = validationResult.getItems();
+
+    // Logs warnings if any.
+    items.stream()
+        .filter(v -> v.getValidation().getLevel().equals(WARN))
+        .forEach(v -> LOGGER.warn(validationResultItemToString(v)));
+
+    // If there are any errors, throws.
+    final boolean hasErrors = items.stream().anyMatch(v -> v.getValidation().getLevel().equals(ERROR));
+    if (hasErrors) {
+      throw new ConfigurationException(createStaticMessage(validationResult.getItems()
+          .stream()
+          .map(ArtifactAstUtils::validationResultItemToString)
+          .collect(joining(lineSeparator()))));
+    }
+  }
+
+  private static String validationResultItemToString(ValidationResultItem v) {
+    return v.getComponents().stream()
+        .map(component -> component.getMetadata().getFileName().orElse("unknown") + ":"
+            + component.getMetadata().getStartLine().orElse(-1))
+        .collect(joining("; ", "[", "]")) + ": " + v.getMessage();
+  }
+
   private static ArtifactAst doParseArtifactIntoAst(String[] configResources,
                                                     AstXmlParserSupplier parserSupplier,
                                                     Set<ExtensionModel> extensions,
-                                                    boolean disableValidations) {
-    try {
-      return parserSupplier.getParser(extensions, disableValidations).parse(loadConfigResources(configResources));
-    } catch (Exception e) {
-      throw new MuleRuntimeException(e);
+                                                    boolean disableValidations)
+      throws ConfigurationException {
+    return parserSupplier.getParser(extensions, disableValidations).parse(loadConfigResources(configResources));
+  }
+
+  private static boolean definesOwnExtensionModel(ArtifactAst artifactAst) {
+    ArtifactType artifactType = artifactAst.getArtifactType();
+    switch (artifactType) {
+      case APPLICATION:
+        return artifactAst.topLevelComponentsStream()
+            .anyMatch(component -> REUSABLE_COMPONENT_TYPES.contains(component.getComponentType()));
+      case MULE_EXTENSION:
+        // TODO W-11796932: Try to unify this one with the case above by changing the Mule Extensions DSL
```

Comment:
The problem is that, at this point, we know the artifact is being parsed as a `MULE_EXTENSION`, but we actually don't know yet what it has inside.
It may be something that is not really an extension (an app's XML for example), because someone got really confused and gave it the wrong XML.
If we remove this check here we will assume it is an extension and will fail horribly before being able to reach the proper validation.
The proper validation is performed over the complete AST, not the partial one. So the artifact-local `ExtensionModel` generation is not protected against malformed artifacts, that is why we need to skip it.

Reply:
I just added a new tests that fails if we change this to always true. This way we are covered for when we do the TODO.